### ISSUE TITLE
Crash in Reader comment containing multiple gifs

### ISSUE THREAD

**Issue by frosty** [wordpress-mobile/WordPress-iOS#10745]
I'm seeing a crash in the Reader's comments view when viewing a comment that contains multiple gifs. The comment that triggers the crash for me is `p73Wbz-ja-p2#comment-5396`. The crash only appears to be happening on device – on my iPhone X on iOS 12.1.2 it happens every time.

Exception is:

```
*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** NSRunStorage (0x28008aa60), _replaceElements(): replaced range {84, 972} extends beyond current run storage size 216.'
```

Stacktrace:
```
#0	0x00000001c0931a08 in objc_exception_throw ()
#1	0x00000001c1667594 in +[NSException raise:format:] ()
#2	0x00000001cbddac20 in _replaceElements ()
#3	0x00000001cbd7f0dc in -[NSLayoutManager(NSPrivate) _invalidateGlyphsForExtendedCharacterRange:changeInLength:includeBlocks:] ()
#4	0x00000001cbd64118 in -[NSLayoutManager(NSPrivate) _fillLayoutHoleForCharacterRange:desiredNumberOfLines:isSoft:] ()
#5	0x00000001cbd6837c in -[NSLayoutManager(NSPrivate) _fillLayoutHoleAtIndex:desiredNumberOfLines:] ()
#6	0x00000001cbd70178 in -[NSLayoutManager(NSPrivate) _markSelfAsDirtyForBackgroundLayout:] ()
#7	0x00000001cbd6a9dc in -[NSLayoutManager(NSPrivate) _invalidateLayoutForExtendedCharacterRange:isSoft:invalidateUsage:] ()
#8	0x00000001cbd6ac3c in -[NSLayoutManager textStorage:edited:range:changeInLength:invalidatedRange:] ()
#9	0x00000001cbd6ab04 in -[NSTextStorage _notifyEdited:range:changeInLength:invalidatedRange:] ()
#10	0x00000001cbd683ec in -[NSTextStorage endEditing] ()
#11	0x00000001eec5acf4 in -[UITextView setAttributedText:] ()
#12	0x00000001048ef364 in WPRichContentView.attributedText.setter ()
#13	0x00000001048ef264 in @objc WPRichContentView.attributedText.setter ()
#14	0x00000001048eec80 in WPRichContentView.content.setter at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:128
#15	0x0000000104f863bc in ReaderCommentCell.configureText() at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Reader/ReaderCommentCell.swift:185
#16	0x0000000104f84c14 in ReaderCommentCell.configureCell(comment:) at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Reader/ReaderCommentCell.swift:127
#17	0x0000000104f84c90 in @objc ReaderCommentCell.configureCell(comment:) ()
#18	0x000000010468a350 in -[ReaderCommentsViewController configureCell:atIndexPath:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m:899
#19	0x000000010468a790 in -[ReaderCommentsViewController tableView:cellForRowAtIndexPath:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/ViewRelated/Reader/ReaderCommentsViewController.m:919
#20	0x000000010463f274 in -[WPTableViewHandler tableView:cellForRowAtIndexPath:] at /Users/frosty/src/WordPress-iOS/WordPress/Classes/Utility/WPTableViewHandler.m:345
#21	0x00000001eeb58de4 in -[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:] ()
```

Exception is happening here: https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift#L128

**Comment by loremattei** [comment]
It looks very close to 59332fd4be077a4dcca33e66-fabric.

**Comment by rachelmcr** [comment]
I'm experiencing this crash when I scroll quickly through the Reader comments view for a specific post (internal ref: p9OQAC-6B-p2). There aren't any GIFs, just PNGs — the crash occurs if I scroll past some of the PNGs before they finish loading. Scrolling slowly and letting the images load doesn't cause any crash.

I saw this crash in WPiOS 11.7.0.20190125 beta on my iPhone XS, iOS 12.1.3.

**Comment by rachelmcr** [comment]
I just got the same crash again on a different Reader comment on a private site. In this case there are no images at all in the comment and it happens every time, no exceptions so far. As soon as I get to where a specific comment should appear on screen, the app crashes. (Internal ref: pb6Nl-cWd-p2 #comment-89924)

Seen on WPiOS 11.9.0.20190226 on iPhone XS, iOS 12.1.4.

**Comment by rfaile313** [comment]
user report 1869386-z

**Comment by rachelmcr** [comment]
@rfaile313 Thanks for sharing the report. Could you include more details about what the user is experiencing? (For example, does it happen on a comment with multiple GIFs, as in the original report, or another case?)

**Comment by rfaile313** [comment]
I've asked for clarification. Here are the logs are well if that's helpful: https://wp.me/p9xo7R-4J

Link is to an internal blog since it contains sensitive information.

**Comment by koke** [comment]
I can consistently reproduce this on a specific internal post as well. I've seen it on a couple different comments, and I think it depends on how fast I scroll. In #11145 @aerych suggests it's about cell reuse, so maybe it has to do with the cell being replaced and not what's in the new comment.

**Comment by etoledom** [comment]
With this comment shared by @rachelmcr (Internal ref: pb6Nl-cWd-p2 #comment-89924), I was able to reproduce it quite constantly, independent of the speed of scrolling.

Even with a very slow scrolling, as soon as that comment appears, the app will crash.

I agree with @koke that it seems more a cell reuse issue, not just the content of the cell, but the content of the comment that was using the same cell before might affect too.

Sometimes scrolling fast makes UIKit to reuse a different cell, and that might explain why sometimes it is dependent on speed.

**Comment by rezzap** [comment]
We have user reporting similar behaviour on a post where they have a lot of content. I was able to get the app to crash when viewing comments on this post in the reader: https://justmakeitadouble.wordpress.com/2019/04/11/to-jess/

Reported in 1957880-zen

I also experience this consistently in the app myself. An example post where I notice this is (Internal ref: p9aQEi-x2-p2) This seems to happen exactly as I scroll over the comment with an image included (jpeg). Replicated on multiple devices. 

Crash logs indicate this error:
```
Fatal Exception: NSRangeException
*** NSRunStorage (0x2826ddfe0), _replaceElements(): replaced range {140, 669} extends beyond current run storage size 269.
```

Seen on WPiOS 12.1.2 on iPhone X, iOS 12.2.

**Comment by thehenrybyrd** [comment]
Per @rezzap's report I was also able to reproduce this with a particular comment
(Internal ref: p9aQEi-x2 #comment-1770), and found both events under 59332fd4be077a4dcca33e66-fabric.

**Comment by thehenrybyrd** [comment]
The user from the above ticket (1957880-zen) came back to clarify that this happens frequently, but not every time, with this particular site they follow. It definitely sounds like this happens most when there's a lot of images or GIFs in the post or comments.

**Comment by aerych** [comment]
This is potentially resolved with an unrelated patch that @jklausa is working on: #11401. In his patch the cell recreates its internal text view vs changing its contents.  I'm optimistic this should avoid the crash. 🤞

**Comment by aerych** [comment]
Call me paranoid, but reopening and self-assigning while we test the beta.

**Comment by aerych** [comment]
We had one hitch which was addressed (hopefully) in #11698. 
Closing :)
